webui/react/src/components/Markdown.tsx

```diff
@@ -0,0 +1,53 @@
+import { Tabs } from 'antd';
+import { default as MarkdownViewer } from 'markdown-to-jsx';
+import React from 'react';
+
+import css from './Markdown.module.scss';
+import Spinner from './Spinner';
+
+const { TabPane } = Tabs;
+const MonacoEditor = React.lazy(() => import('components/MonacoEditor'));
+
+interface Props {
+  editing?: boolean;
+  height?: string;
+  markdown: string;
+  onChange?: (editedMarkdown: string) => void;
+}
+
+const Markdown: React.FC<Props> = ({ editing=false, height='80vh', markdown, onChange }: Props) => {
+  return (
+    editing ?<Tabs>
+      <TabPane key="1" tab="Edit">
+        <React.Suspense
+          fallback={<div><Spinner tip="Loading text editor..." /></div>}>
+          <MonacoEditor
+            defaultValue={markdown}
+            height={height}
+            language="markdown"
+            options={{
+              wordWrap: 'on',
+              wrappingIndent: 'indent',
+            }}
+            width="100%"
+            onChange={onChange} />
+        </React.Suspense>
+      </TabPane>
+      <TabPane key="2" tab="Preview">
+        <div className={css.base} style={{ height }}>
+          <MarkdownViewer options={{ disableParsingRawHTML: true }}>
+            {markdown}
+          </MarkdownViewer>
+        </div>
+      </TabPane>
+    </Tabs> :
+      <div className={css.base} style={{ height, overflow: 'auto' }}>
```

Comment:
**suggestion:** having written some interaction tests recently, one thing that came up that I think impacts the JSX design is that it might be better to have one parent div in the reusable component that remains consistent. Meaning, taking this setup for example, sometimes the top level is a `<Tab>`, and other times it's a generic `<div>` where the reference to the component element changes. I think that behavior is ok for pages, but for reusable components, the following pattern would be preferred (for testing purposes too).

```
return (
  <div aria-label="markdown-editor" className={css.base} style={...}>
    {editing ? (
      <Tab />
    ) : (
      <MarkdownViewer />
    )}
  </div>
);
```

The reason for why this is nice for interaction testing and accessibility is that it will always render a DOM element and when we start using more and more aria-labels for accessibility compatibility, it will be a single place to apply it along with styling. Even if the base element won't show up via CSS, the reference to the component is stable and reliable.